### ISSUE TITLE
[FEATURE REQUEST]Calculate topK on GPU with k-selection algos

### ISSUE THREAD

**Issue by lagoon0o0** [tensorflow/tensorflow#5719]
Hi,
I am currently using [tf.nn.top_k](https://www.tensorflow.org/versions/r0.11/api_docs/python/nn.html#top_k) in my code. However, I found this operation is calculated on CPU and it is quite slow.
So I am wondering if it is possible to calculate topK on GPU?

Here are the papers and cuda codes about k-selection on GPU:
http://www.math.grin.edu/~blanchaj/Research/ABGS_KSelection.pdf
https://code.google.com/p/ggks/

Hope someday this feature will be added to tensorflow.

**Comment by eth-n** [comment]
Hello, I'd love to contribute by working on an implementation of this. I'm considering the in-place radix k-select, pivot to cut problem size, then sort, or perhaps a full sort if that turns out to be quicker.

@lagoon0o0, do you have a history of runtimes for certain size/shape tensors for this algorithm you'd like to see improvement on?

I'm fairly new to TensorFlow, and would appreciate some insight into the shapes that tensors take in production applications. Thanks!

**Comment by christopher5106** [comment]
+1

**Comment by NaveenAri** [comment]
This mixture of experts paper (https://arxiv.org/pdf/1701.06538.pdf) from google brain seems to heavily use topk, so maybe we can hope for an update soon?

**Comment by ed-alertedh** [comment]
I'm also interested in implementing this as it's currently bottlenecking me. If there is already an implementation internally at Google brain (seems likely) or if @eth-n has already started then there probably isn't much point though.

**Comment by ed-alertedh** [comment]
For those interested, I've started attempting an implementation of this. Unfortunately I'm having to learn CUDA as I go, so it's slow going. I figure if I make something suboptimal that avoids D2H/H2D transfers then that is still a good starting point.

The GGKS code would need some modification so it couldn't just be placed in the third_party folder. That presumably would mean we'd need to get the original authors to sign the CLA in addition to myself - perhaps @prb12 or @aselle can confirm my assumption there? Given that the paper referenced above has enough detail to implement it from scratch, I've decided to go that route for now (making sure not to look at the original code to avoid plagiarism).

The fact that this op also requires a sort at the end makes things a little messier. That is probably something that could best be achieved using a third_party lib such as Nvidia's [thrust](https://github.com/thrust/thrust) (licensed under Apache 2.0). This could also be used to make a dedicated sort op (see issue #288). I'll focus on the selection algorithm for now, perhaps others might have some input?

**Comment by ed-alertedh** [comment]
Update: I have a working GPU implementation but based on some very informal benchmarking, performance is only really better for huge tensors. Still need to clean it up a lot, but after that I might upload it to see if anyone has any tips on how to improve it.

**Comment by khaotik** [comment]
@ed-alertedh May I see your code? I'm doing something similar, for Theano. Thanks.

**Comment by andrei-pokrovsky** [comment]
You could use a GPU cuda radix sort sample that comes with standard CUDA distribution as a starting point for implementing topK (even though there's probably a simpler way to do it in CUDA)

**Comment by khaotik** [comment]
I've ported PyTorch implementation for Theano at [here](https://github.com/Theano/Theano/pull/5959), which uses radix selection. I mostly works but still need some cleanup.

**Comment by ed-alertedh** [comment]
Well, it looks like we got what we wanted! @ebrevdo looks to have implemented this and it's in 1.3.0rc0!
Interestingly, it's a heap-based algorithm: [code here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/topk_op_gpu.cu.cc)

I'm a little sad that I didn't manage to release my code in time due to unforeseen "red tape" at work. But OTOH, I'm a newbie with CUDA and my thrust-based implementation wasn't really that great anyway :P

**Comment by yliapis** [comment]
@ebrevdo @ed-alertedh I tried running the new top_k operator as an intermediate layer on the gpu, and am still getting cpu-like speeds. Any idea what could be going on? I logged the device placement, and it does seem like the operator is still getting placed on the gpu. 

I tested a script with the operator in between two convolutional layers on both 1.3.0-rc2 (gpu) and 1.2.1 (cpu), and the cpu version was faster. Any idea what the bottleneck could be?

**Comment by ed-alertedh** [comment]
@yliapis What shape are the tensors you're running top_k on? They're probably just too small to benefit from the GPU op. I've just been using the CPU op, although in my use-case I've found k=1 is often sufficient so in that case I can just use the max op instead.

On another note, I tried to run the new GPU op on a tensor of shape [1, 10,000] and got some obscure CUDA error. Haven't had time to track down possible cause or file a bug report yet. I know with thrust I had a lot of fun trying to integrate with TF's memory allocator - as far as I could tell, it doesn't allow kernels to dynamically allocate memory so if CUB is using malloc inside kernels it could just be running out of memory because TF has grabbed it all...

**Comment by yliapis** [comment]
@ed-alertedh The tensors I am using the op on are generally [256, 20] up to [3600, 20]. Note that I am using tf.map_fn to map a tensorflow operator using top_k to a batch (32-128 usually) of matrices, so the top_k function is getting called batch_sz times by map_fn. I haven't been getting any cuda errors, but the performance makes me question whether the gpu is being properly utilized.

**Comment by ed-alertedh** [comment]
I guess it just isn't scaling well for that number of rows. Not really what you'd expect given it should be an "embarassingly parallel" problem, but maybe you just need to throw a huge number of rows at it to start seeing the benefit. Perhaps concatenating the matrices rather than using map_fn would change performance? Would likely eat up too much memory though....

Hard to say for sure because the benchmarks were all run with 128 rows.  Your case of [256,20] is roughly comparable to the benchmarks on [128,10] and [128,100] which all ran faster on CPU. In case you hadn't found them, benchmark results here: https://github.com/tensorflow/tensorflow/commit/0b5cce367cf95a5d7fbb6b037e7a8646f6c47b70

**Comment by yliapis** [comment]
@ed-alertedh @ebrevdo I tried top_k by itself on the gpu and the cpu with m=10**6, n=20. The cpu version was twice as fast. Are the rows being sorted in parallel?

**Comment by ed-alertedh** [comment]
It does seem counter-intuitive that performance would be that much worse than CPU on that many rows...
If it helps, I was using the following to force a DtoH/HtoD transfer when benchmarking the CPU op so I could better compare CUDA overheads versus the overhead of transferring to CPU:
```
from tensorflow.python.platform import benchmark
from tensorflow.python.platform import test
import tensorflow as tf

class TopKBenchmark(test.Benchmark):
    def benchmark_topk_gpu(self, dims, k, sorted=False):
        with tf.device('/gpu:0'):
            test_data = np.random.uniform(size=dims)
            data = tf.constant(test_data, dtype=tf.float64)
            k_in = tf.constant(k)

        with tf.device('/gpu:0'):
            op_under_test = tf.nn.top_k(data, k_in, sorted=sorted)

        # Creates a session with log_device_placement set to True.
        conf = tf.ConfigProto(log_device_placement=True, allow_soft_placement=False)
        # Don't perform optimizations for tests so we don't inadvertently run
        # gpu ops on cpu
        conf.graph_options.optimizer_options.opt_level = -1
        sess = tf.Session(config=conf)
        with sess:
            return self.run_op_benchmark(
                sess, op_under_test, min_iters=1000, store_trace=False, name="op_benchmark")

    def benchmark_topk_cpu(self, dims, k, sorted=False)):
        with tf.device('/gpu:0'):
            test_data = np.random.uniform(size=dims)
            data = tf.constant(test_data, dtype=tf.float64)
            k_in = tf.constant(k)

        with tf.device('/cpu:0'):
            op_under_test = tf.nn.top_k(data, k_in, sorted=sorted)

        # Creates a session with log_device_placement set to True.
        conf = tf.ConfigProto(log_device_placement=True, allow_soft_placement=False)
        # Don't perform optimizations for tests so we don't inadvertently run
        # gpu ops on cpu
        conf.graph_options.optimizer_options.opt_level = -1
        sess = tf.Session(config=conf)
        with sess:
            return self.run_op_benchmark(
                sess, op_under_test, min_iters=1000, store_trace=False, name="op_benchmark")


bm = TopKBenchmark()
dims = (1e6, 20)
k=5
cpu_res = bm.benchmark_topk_cpu(dims, k)
gpu_res = bm.benchmark_topk_gpu(dims, k)
print('***** CPU TIME:')
print(cpu_res)
print('***** GPU TIME:')
print(gpu_res)
```
I won't pretend to fully understand the current implementation since I've only skim-read it, but the sort implementation looks like it should be parallel on the rows: https://nvlabs.github.io/cub/structcub_1_1_device_segmented_radix_sort.html

**Comment by BlackHC** [comment]
You don't have that many rows to be fair, you do have a big batch dimension, however.

Try maybe `(20,10**6)`, so 20 batches a 1 million rows. The GPU topk implementation launches one thread block per batch, so with the original `(10**6,20)`, it will perform a topk on 20 elements over a 1 million batches. Probably not what you want. (Or at least I don't think that is the general use-case).

Thanks,
 Andreas

**Comment by ed-alertedh** [comment]
Yeah, I think it is just inherent in how this problem scales to GPU - "small" inputs don't overcome the fixed costs of kernel launch. I hope it doesn't come across that we're criticising your implementation! Just figuring out what the scaling behaviour is.

For the sake of clarity, in my above statements I was using the convention [rows, columns] whereas @BlackHC seems to be using the convention [batch size, rows].

If it only launches one thread block per batch (=1024 threads on most devices), wouldn't that mean the implementation scales well to large batch sizes and doesn't scale well to large numbers of rows? (using your convention [batch size, rows]).

**Comment by ebrevdo** [comment]
There are two different implementations, depending on the input sizes.
Code is here
<https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/topk_op_gpu.cu.cc#L545>.
We use cub sort followed by copy of the top k for one range of parameters
and we use @BlackHC's heap for another range.

There may actually be a bug where we decide to use cub for num_cols <=
1000; but perhaps that should have been use cub for num_cols >= 1000
instead.

**Comment by ed-alertedh** [comment]
Ah right, forgot about that detail. So for [20, 10 ** 6] the heap will be used whereas for [10**6, 20] the cub sort will be used.

Using cub for `num_cols<=1000` seems intuitive to me - use a heap for larger problem sizes to save the effort of sorting the whole thing. But once k gets large enough it approaches a heapsort, hence why you'd switch to a more performant sorting algo. For non-parallel versions of the algorithms I believe the complexity comparison is O(n + k log(n)) for heap versus O(n log(n)) for sort. Obviously you can't directly apply that comparison to the parallel algorithms, but you'd expect the heap to become faster at some point where  k << n.

I think my previous comment still makes sense - the heap implementation currently shards into a maximum of 1024 threads per batch all in one thread block. Each thread builds its own heap on a subset of the data. If the shards were split over multiple blocks you could increase the parallelism. But that would require a parallel merge to gain any benefit and it would be merging between blocks which means you risk getting bottlenecked by global memory. I can see why you'd want to keep it within a single block. If you had large enough num_cols, it would eventually become worth it to use multiple blocks.

**Comment by nitaidean** [comment]
I tried using top_k on multiple GPU's using TensorFlow 1.3.0, and it failed with this message:

**Cannot assign a device for operation 'gradients/TopKV2_grad/sub_1': Could not satisfy explicit device specification '/device:GPU:0' because no supported kernel for GPU devices is available.**

This suggests to me that top_k still does not have GPU kernel support, and hence must still run on the CPU. But this thread mentioned that the problem had been solved by @ebrevdo back in 1.3.0rc0. Did I misunderstand?

**Comment by ebrevdo** [comment]
Does it work in tf nightlies?

**Comment by nitaidean** [comment]
Just tried it - nope. Same error. Is it supposed to work?

**Comment by ed-alertedh** [comment]
Peanut gallery chiming in: What datatype is the tensor you're trying to run top_k on? Practically none of the GPU ops are defined for int types (see #9506)

**Comment by nitaidean** [comment]
Yeah, I read that, but no - it's a tensor of floats.

**Comment by ed-alertedh** [comment]
After looking more closely at the error message, my guess would be that one or more of the ops in the gradient definition [here](https://github.com/tensorflow/tensorflow/blob/c1832374a6a311837f2738e08a9bff8480c8ede4/tensorflow/python/ops/nn_grad.py#L903) can't run on GPU due to the limitation on ints

**Comment by nitaidean** [comment]
Ah, so the problem was never fixed in the first place then? Has anyone else succeeded in getting top_k to run on GPU under any scenario?

**Comment by ed-alertedh** [comment]
It will still be useful in scenarios where you don't need the gradient. As a workaround I'd suggest enabling soft placement so the gradient ops can be placed on CPU.

I did get some CUDA kernel launch errors with the new implementation but that would be a totally separate issue. Still haven't had time to prepare a proper bug report (it could have been fixed in the meantime anyway)

**Comment by ebrevdo** [comment]
It should work in the nightlies, if you have a full stack trace and minimal
graph/code to replicate that would be helpful.

**Comment by ed-alertedh** [comment]
OK, looks like you've made significant changes since I last tried it, but I'm still seeing the same behaviour on tf-nightly-gpu==1.4.0.dev20171013. Info below but let me know if I can gather any more - sorry for not reporting earlier!

Minimal working example for my machine:
```
import tensorflow as tf
import numpy as np

FLAGS = tf.flags.FLAGS

tf.flags.DEFINE_integer('m', 1, 'batch size')
tf.flags.DEFINE_integer('n', 1, 'data length')
tf.flags.DEFINE_integer('k', 1, 'pick the top k')

dims = (FLAGS.m, FLAGS.n)

with tf.device('/gpu:0'):
    test_data = np.random.uniform(size=dims)
    data = tf.constant(test_data, dtype=tf.float64)
    k_in = tf.constant(FLAGS.k)
    op_under_test = tf.nn.top_k(data, k_in, sorted=False)

conf = tf.ConfigProto(log_device_placement=True, allow_soft_placement=False)
# Don't perform optimizations for tests so we don't inadvertently run
# gpu ops on cpu
conf.graph_options.optimizer_options.opt_level = -1
with tf.Session(config=conf) as sess:
    sess.run(op_under_test)
```

Error log for input of size [1, 6140]
[error_log.txt](https://github.com/tensorflow/tensorflow/files/1386529/error_log.txt)
Normal execution for input of size [1,6139]
[normal_execution.txt](https://github.com/tensorflow/tensorflow/files/1386530/normal_execution.txt)

**Comment by carusyte** [comment]
In a [Differentiable Neural Computer (DNC)](https://github.com/deepmind/dnc) model, top_k is used in memory write access, which is in the RNN while loop, and this degrades performance tremendously. I'm looking forward to the implementation of GPU top_k, too.

**Comment by carusyte** [comment]
I submitted a PR #21436 which attempts to workaround the issue of gradient execution on CPU, waiting for code review. If anybody is still interested...

**Comment by ebrevdo** [comment]
Thanks!

**Comment by ed-alertedh** [comment]
FYI the error I reported above was addressed in https://github.com/tensorflow/tensorflow/commit/6bcc00668094be8daf8465b8689bbed5ab285b2d and running the script I posted in the same windows env with tf 1.10 no longer throws an error.
